{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Projet3.ipynb\n",
      "regulatory_seq_PHO.fasta\n",
      "regulatory_seqs_GAL.fasta\n",
      "regulatory_seqs_MET.fasta\n",
      "yeast_s_cerevisae_genomic_chr1-4.fna\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import projet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:\n",
      "UnicodeDecodeError while processing traceback.\n",
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'nucleobase_undetermine' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "projet.read_fasta(\"regulatory_seq_PHO.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import math\n",
    "import numpy as np\n",
    "import copy\n",
    "\n",
    "#Définition des nucléotides\n",
    "nucleotide = {'A':0,'C':1,'G':2,'T':3}\n",
    "nucleotide_inverse = {0:'A',1:'C',2:'G',3:'T'}\n",
    "nucleotide_indetermine = {'A':0,'C':1,'G':2,'T':3,'N':-1}\n",
    "\n",
    "def decode_sequence(sequence):\n",
    "    inv_nucleotide = {v:k for k, v in nucleotide_indetermine.items()}\n",
    "    to_str = \"\"\n",
    "    for i in sequence:\n",
    "        if(i in inv_nucleotide):\n",
    "            to_str += inv_nucleotide[i]\n",
    "        else:\n",
    "            to_str += 'N'\n",
    "    return to_str\n",
    "    \n",
    "\n",
    "def encode_sequence(string):\n",
    "    to_list = []\n",
    "    for base in string:\n",
    "        if(base in nucleotide_indetermine):\n",
    "            to_list.append(nucleotide_indetermine[base])\n",
    "    return to_list\n",
    "\n",
    "def read_fasta(fasta_filepath):\n",
    "    fasta_file = io.open(fasta_filepath, 'r')\n",
    "    current_sequence = \"\"\n",
    "    sequences_dict = {}\n",
    "    for line in fasta_file.readlines():\n",
    "        if(line[0] == '>'):\n",
    "            current_sequence = line\n",
    "            sequences_dict[line] = []\n",
    "        else:\n",
    "            for nucl in line:\n",
    "                if(nucl in nucleotide_indetermine):\n",
    "                    sequences_dict[current_sequence].append(nucleotide_indetermine[nucl])\n",
    "    \n",
    "\n",
    "    return sequences_dict\n",
    "    \n",
    "def nucleotide_count(sequence):\n",
    "    count = [0 for k in nucleotide]\n",
    "    for nucl in sequence:\n",
    "        if(nucl >= 0):\n",
    "            count[nucl] += 1\n",
    "    return count\n",
    "\n",
    "def nucleotide_frequency(sequence):\n",
    "    count = [0 for k in nucleotide]\n",
    "    n_nucl = 0.\n",
    "    for nucl in sequence:\n",
    "        if(nucl >= 0):\n",
    "            count[nucl] += 1\n",
    "            n_nucl += 1.\n",
    "    return count/(np.sum(count))\n",
    "\n",
    "def log_proba_sequence(liste_freq):\n",
    "    proba = 1\n",
    "    for i in liste_freq:\n",
    "        proba *= i\n",
    "    return math.log(proba)\n",
    "\n",
    "def count_word(k, sequence):\n",
    "    res = {}\n",
    "    \n",
    "    for i in range(len(sequence)-k+1):\n",
    "        mot = sequence[i:i+k]\n",
    "        string = \"\"\n",
    "        for s in mot:\n",
    "            string += nucleotide_inverse[s]\n",
    "        if string in res.keys():\n",
    "            res[string] += 1\n",
    "        else:\n",
    "            res[string] = 1\n",
    "            \n",
    "    return res\n",
    "        \n",
    "    \n",
    "def probaCondi(frequences, mot):\n",
    "    reverseMot = copy(mot)\n",
    "    reverseMot.reverse()\n",
    "    characters = list(reverseMot)\n",
    "    proba = 1\n",
    "    for i in range(0, len(characters)):\n",
    "        proba *= frequences[nucleotide[characters[0]]]\n",
    "    return proba\n",
    "    \n",
    "def comptage_attendu(sequence, frequences, k, l):\n",
    "    #UTILISER LOGPROBA\n",
    "    result = {}\n",
    "    mots = count_word(k, sequence).keys()\n",
    "    for mot in mots:\n",
    "        result[mot] = probaCondi(frequences, mot)*l\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = read_fasta(\"regulatory_seq_PHO.fasta\")\n",
    "nom = list(sequence.keys())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">PHO81\tPHO81; upstream from -800 to -1; size: 800; feature type:cds; location: Saccharomyces_cerevisiae:NC_001139.8:958214:959013:R\n",
      " [251, 145, 145, 259]\n",
      "[0.31375 0.18125 0.18125 0.32375]\n",
      "-5.702698424999239\n",
      "\n",
      "\n",
      "\n",
      ">PHO11\tPHO11; upstream from -800 to -1; size: 800; feature type:cds; location: Saccharomyces_cerevisiae:NC_001133.7:224651:225450:D\n",
      " [289, 145, 139, 227]\n",
      "[0.36125 0.18125 0.17375 0.28375]\n",
      "-5.735862529526608\n",
      "\n",
      "\n",
      "\n",
      ">PHO84\tPHO84; upstream from -800 to -1; size: 800; feature type:cds; location: Saccharomyces_cerevisiae:NC_001145.2:25802:26601:R\n",
      " [269, 182, 136, 213]\n",
      "[0.33625 0.2275  0.17    0.26625]\n",
      "-5.665781792547597\n",
      "\n",
      "\n",
      "\n",
      ">PHO8\tPHO8; upstream from -800 to -1; size: 800; feature type:cds; location: Saccharomyces_cerevisiae:NC_001136.8:1420243:1421042:R\n",
      " [267, 162, 198, 173]\n",
      "[0.33375 0.2025  0.2475  0.21625]\n",
      "-5.6220432918467615\n",
      "\n",
      "\n",
      "\n",
      ">PHO5\tPHO5; upstream from -800 to -1; size: 800; feature type:cds; location: Saccharomyces_cerevisiae:NC_001134.7:430946:431745:R\n",
      " [255, 160, 158, 227]\n",
      "[0.31875 0.2     0.1975  0.28375]\n",
      "-5.634464499771086\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "complet = []\n",
    "for k in nom:\n",
    "    complet = complet + sequence[k]\n",
    "    print(k, nucleotide_count(sequence[k]))\n",
    "    print(nucleotide_frequency(sequence[k]))\n",
    "    print(log_proba_sequence(nucleotide_frequency(sequence[k])))\n",
    "    print(\"\\n\\n\")\n",
    "sequence[\"complet\"] = complet\n",
    "sequenceLength = nucleotide_count(sequence[\"complet\"])\n",
    "letterFrequency = nucleotide_frequency(sequence[\"complet\"])\n",
    "sequenceLog = log_proba_sequence(nucleotide_frequency(sequence[\"complet\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AAAA': 12,\n",
       " 'AAAC': 3,\n",
       " 'AAAG': 3,\n",
       " 'AAAT': 8,\n",
       " 'AACA': 5,\n",
       " 'AACC': 4,\n",
       " 'AACG': 6,\n",
       " 'AACT': 4,\n",
       " 'AAGA': 6,\n",
       " 'AAGC': 1,\n",
       " 'AAGG': 2,\n",
       " 'AAGT': 5,\n",
       " 'AATA': 5,\n",
       " 'AATC': 5,\n",
       " 'AATG': 3,\n",
       " 'AATT': 11,\n",
       " 'ACAA': 9,\n",
       " 'ACAG': 3,\n",
       " 'ACAT': 5,\n",
       " 'ACCA': 3,\n",
       " 'ACCC': 1,\n",
       " 'ACCG': 1,\n",
       " 'ACCT': 2,\n",
       " 'ACGA': 1,\n",
       " 'ACGC': 3,\n",
       " 'ACGT': 7,\n",
       " 'ACTA': 7,\n",
       " 'ACTG': 2,\n",
       " 'ACTT': 2,\n",
       " 'AGAA': 8,\n",
       " 'AGAC': 2,\n",
       " 'AGAG': 4,\n",
       " 'AGAT': 3,\n",
       " 'AGCA': 6,\n",
       " 'AGCC': 1,\n",
       " 'AGCG': 1,\n",
       " 'AGCT': 4,\n",
       " 'AGGC': 2,\n",
       " 'AGGG': 2,\n",
       " 'AGGT': 5,\n",
       " 'AGTA': 2,\n",
       " 'AGTC': 2,\n",
       " 'AGTG': 1,\n",
       " 'AGTT': 5,\n",
       " 'ATAA': 8,\n",
       " 'ATAC': 3,\n",
       " 'ATAG': 5,\n",
       " 'ATAT': 6,\n",
       " 'ATCA': 7,\n",
       " 'ATCG': 1,\n",
       " 'ATCT': 5,\n",
       " 'ATGA': 3,\n",
       " 'ATGC': 2,\n",
       " 'ATGG': 3,\n",
       " 'ATGT': 3,\n",
       " 'ATTA': 8,\n",
       " 'ATTC': 3,\n",
       " 'ATTG': 7,\n",
       " 'ATTT': 9,\n",
       " 'CAAA': 6,\n",
       " 'CAAC': 7,\n",
       " 'CAAG': 8,\n",
       " 'CAAT': 4,\n",
       " 'CACA': 3,\n",
       " 'CACG': 2,\n",
       " 'CACT': 3,\n",
       " 'CAGA': 1,\n",
       " 'CAGC': 4,\n",
       " 'CAGG': 2,\n",
       " 'CAGT': 2,\n",
       " 'CATA': 4,\n",
       " 'CATC': 3,\n",
       " 'CATG': 2,\n",
       " 'CATT': 3,\n",
       " 'CCAA': 1,\n",
       " 'CCAC': 2,\n",
       " 'CCAT': 1,\n",
       " 'CCCT': 3,\n",
       " 'CCGC': 1,\n",
       " 'CCGT': 2,\n",
       " 'CCTA': 4,\n",
       " 'CCTC': 2,\n",
       " 'CCTT': 5,\n",
       " 'CGAA': 1,\n",
       " 'CGAG': 2,\n",
       " 'CGAT': 2,\n",
       " 'CGCC': 1,\n",
       " 'CGCG': 2,\n",
       " 'CGCT': 2,\n",
       " 'CGGT': 2,\n",
       " 'CGTA': 3,\n",
       " 'CGTC': 1,\n",
       " 'CGTG': 5,\n",
       " 'CGTT': 2,\n",
       " 'CTAA': 3,\n",
       " 'CTAC': 4,\n",
       " 'CTAG': 3,\n",
       " 'CTAT': 7,\n",
       " 'CTCA': 2,\n",
       " 'CTCC': 1,\n",
       " 'CTCG': 1,\n",
       " 'CTCT': 1,\n",
       " 'CTGC': 4,\n",
       " 'CTTA': 3,\n",
       " 'CTTC': 4,\n",
       " 'CTTG': 4,\n",
       " 'CTTT': 9,\n",
       " 'GAAA': 5,\n",
       " 'GAAC': 3,\n",
       " 'GAAT': 6,\n",
       " 'GACA': 3,\n",
       " 'GAGA': 5,\n",
       " 'GAGC': 4,\n",
       " 'GAGG': 4,\n",
       " 'GAGT': 1,\n",
       " 'GATA': 3,\n",
       " 'GATG': 3,\n",
       " 'GATT': 4,\n",
       " 'GCAA': 8,\n",
       " 'GCAC': 4,\n",
       " 'GCAG': 3,\n",
       " 'GCAT': 4,\n",
       " 'GCCC': 1,\n",
       " 'GCCT': 3,\n",
       " 'GCGA': 3,\n",
       " 'GCGC': 1,\n",
       " 'GCGT': 1,\n",
       " 'GCTA': 4,\n",
       " 'GCTC': 2,\n",
       " 'GCTG': 1,\n",
       " 'GCTT': 5,\n",
       " 'GGAA': 2,\n",
       " 'GGAG': 1,\n",
       " 'GGCA': 5,\n",
       " 'GGCG': 1,\n",
       " 'GGCT': 2,\n",
       " 'GGGA': 1,\n",
       " 'GGGC': 1,\n",
       " 'GGGT': 1,\n",
       " 'GGTA': 2,\n",
       " 'GGTC': 1,\n",
       " 'GGTG': 2,\n",
       " 'GGTT': 4,\n",
       " 'GTAA': 1,\n",
       " 'GTAC': 5,\n",
       " 'GTAG': 1,\n",
       " 'GTAT': 2,\n",
       " 'GTCA': 1,\n",
       " 'GTCC': 2,\n",
       " 'GTCT': 2,\n",
       " 'GTGA': 4,\n",
       " 'GTGC': 6,\n",
       " 'GTGG': 2,\n",
       " 'GTGT': 1,\n",
       " 'GTTA': 5,\n",
       " 'GTTC': 1,\n",
       " 'GTTG': 1,\n",
       " 'GTTT': 6,\n",
       " 'TAAA': 2,\n",
       " 'TAAC': 6,\n",
       " 'TAAG': 3,\n",
       " 'TAAT': 6,\n",
       " 'TACA': 6,\n",
       " 'TACC': 3,\n",
       " 'TACG': 3,\n",
       " 'TACT': 4,\n",
       " 'TAGA': 5,\n",
       " 'TAGC': 3,\n",
       " 'TAGG': 1,\n",
       " 'TAGT': 2,\n",
       " 'TATA': 10,\n",
       " 'TATC': 5,\n",
       " 'TATG': 3,\n",
       " 'TATT': 9,\n",
       " 'TCAA': 7,\n",
       " 'TCAC': 2,\n",
       " 'TCAG': 3,\n",
       " 'TCAT': 2,\n",
       " 'TCCA': 1,\n",
       " 'TCCC': 1,\n",
       " 'TCCG': 2,\n",
       " 'TCCT': 3,\n",
       " 'TCGA': 1,\n",
       " 'TCGG': 2,\n",
       " 'TCGT': 1,\n",
       " 'TCTA': 2,\n",
       " 'TCTC': 1,\n",
       " 'TCTG': 1,\n",
       " 'TCTT': 8,\n",
       " 'TGAA': 3,\n",
       " 'TGAC': 1,\n",
       " 'TGAG': 7,\n",
       " 'TGAT': 5,\n",
       " 'TGCA': 9,\n",
       " 'TGCC': 2,\n",
       " 'TGCG': 1,\n",
       " 'TGCT': 4,\n",
       " 'TGGA': 2,\n",
       " 'TGGC': 5,\n",
       " 'TGGG': 1,\n",
       " 'TGGT': 1,\n",
       " 'TGTA': 2,\n",
       " 'TGTC': 1,\n",
       " 'TGTG': 5,\n",
       " 'TGTT': 2,\n",
       " 'TTAA': 5,\n",
       " 'TTAC': 4,\n",
       " 'TTAG': 2,\n",
       " 'TTAT': 12,\n",
       " 'TTCA': 4,\n",
       " 'TTCC': 4,\n",
       " 'TTCG': 2,\n",
       " 'TTCT': 4,\n",
       " 'TTGA': 9,\n",
       " 'TTGC': 4,\n",
       " 'TTGG': 4,\n",
       " 'TTGT': 6,\n",
       " 'TTTA': 7,\n",
       " 'TTTC': 6,\n",
       " 'TTTG': 11,\n",
       " 'TTTT': 16}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_word(4, sequence[nom[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'module' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-103-3be7fb0b91c9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcomptage_attendu\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m0.25\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.25\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.25\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.25\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-102-ebeee74ee87b>\u001b[0m in \u001b[0;36mcomptage_attendu\u001b[0;34m(sequence, frequences, k, l)\u001b[0m\n\u001b[1;32m     94\u001b[0m     \u001b[0mmots\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcount_word\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     95\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mmot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmots\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m         \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmot\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprobaCondi\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfrequences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     97\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-102-ebeee74ee87b>\u001b[0m in \u001b[0;36mprobaCondi\u001b[0;34m(frequences, mot)\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mprobaCondi\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfrequences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 84\u001b[0;31m     \u001b[0mreverseMot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     85\u001b[0m     \u001b[0mreverseMot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m     \u001b[0mcharacters\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreverseMot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'module' object is not callable"
     ]
    }
   ],
   "source": [
    "comptage_attendu([0,1,2,3], [0.25,0.25,0.25,0.25], 2, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
